simtfl/node.py

```diff
@@ -78,12 +105,72 @@ def run(self):
         implementation.
         """
         while True:
-            while len(self.mailbox) > 0:
-                (sender, message) = self.mailbox.popleft()
+            while len(self._mailbox) > 0:
+                (sender, message) = self._mailbox.popleft()
                 print(f"T{self.env.now:5d}: handling  {sender:2d} -> {self.ident:2d}: {message}")
                 yield from self.handle(sender, message)
 
             # This naive implementation is fine because we have no actual
             # concurrency.
-            self.wakeup = self.env.event()
-            yield self.wakeup
+            self._wakeup = self.env.event()
+            yield self._wakeup
+
+
+__all__ = ['PassiveNode', 'SequentialNode']
+
+from simpy import Environment
+import unittest
+
+from .message import PayloadMessage
+from .network import Network
+
+
+class PassiveReceiverTestNode(PassiveNode):
+    def __init__(self):
+        super().__init__()
+        self.received = deque()
+
+    def handle(self, sender, message):
+        self.received.append((sender, message, self.env.now))
+        yield self.env.timeout(3)
+
+
+class SequentialReceiverTestNode(SequentialNode):
+    def __init__(self):
+        super().__init__()
+        self.received = deque()
+
+    def handle(self, sender, message):
+        self.received.append((sender, message, self.env.now))
+        yield self.env.timeout(3)
+
+
+class SenderTestNode(PassiveNode):
+    def run(self):
+        for i in range(3):
+            yield from self.send(0, PayloadMessage(i))
+            yield self.env.timeout(1)
+
+
+class TestFramework(unittest.TestCase):
+    def _test_node(self, receiver_node, expected):
+        network = Network(Environment())
+        network.add_node(receiver_node)
+        network.add_node(SenderTestNode())
+        network.run_all()
+
+        self.assertEqual(list(network.node(0).received), expected)
+
+    def test_passive_node(self):
+        self._test_node(PassiveReceiverTestNode(), [
+            (1, PayloadMessage(0), 1),
+            (1, PayloadMessage(1), 2),
+            (1, PayloadMessage(2), 3),
```

Comment:
Here the messages are being sent at times 0, 1, 2, received at times 1, 2, 3, and handled immediately. Since they are handled concurrently, there is no blocking even though handling each one takes 3 time units.

Reply:
Hm… I'm still confused about how `env.timeout` is used and perhaps how `simpy` works.

When considering [these lines](https://github.com/zcash/simtfl/pull/18/files/92c808219341515ea8d4b66171f7d4a40c0fad37#diff-c427d335353ee0d84c4e3021616ea3ea6c2fd1211dff52d89692958fe2a3fee4R151-R152) quoted here:

```
            yield from self.send(0, PayloadMessage(i))
            yield self.env.timeout(1)
```

My mental model is that the first line will cause `SequentialReceiverTestNode.handle` to be called all within time 0 (after traversing through the mailbox). After this, the sender process pauses for 1 time unit. Doesn't this model 0 network propagation delay?

Am I confused about `simpy` plus this framework's beahvior, or am I confused about what each method call is intended to represent in the model?

Something must be wrong about my understanding of `simpy` + this framework, because the assertion clearly shows the first message received at time 1. Can you clarify this?

Reply:
Node classes have nothing to do with modelling network propagation delay. That's done in `Network.convey`.

This is just saying that the node waits 1 time unit after sending each message. The timeout doesn't represent anything, it's just what this particular node does in this test. The reason why it does it is so that we can distinguish the times of the different messages when checking the expected timing.